// apeman-react-spinner/shim/node/ap_spinner_style.js
/**
 * Style for ApSpinner.
 * @class ApSpinnerStyle
 */

'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _apemanReactStyle = require('apeman-react-style');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

/** @lends ApSpinnerStyle */
var ApSpinnerStyle = function (_Component) {
  _inherits(ApSpinnerStyle, _Component);

  function ApSpinnerStyle() {
    _classCallCheck(this, ApSpinnerStyle);

    return _possibleConstructorReturn(this, Object.getPrototypeOf(ApSpinnerStyle).apply(this, arguments));
  }

  _createClass(ApSpinnerStyle, [{
    key: 'render',
    value: function render() {
      var s = this;
      var props = s.props;

      var _ApSpinnerStyle$style = ApSpinnerStyle.styleData(props);

      var all = _ApSpinnerStyle$style.all;
      var small = _ApSpinnerStyle$style.small;
      var medium = _ApSpinnerStyle$style.medium;
      var large = _ApSpinnerStyle$style.large;


      return _react2.default.createElement(
        _apemanReactStyle.ApStyle,
        { data: Object.assign(all, props.style),
          smallMediaData: small,
          mediumMediaData: medium,
          largeMediaData: large
        },
        props.children
      );
    }
  }]);

  return ApSpinnerStyle;
}(_react.Component);

Object.assign(ApSpinnerStyle, {
  propTypes: {
    type: _react.PropTypes.string,
    style: _react.PropTypes.object
  },
  defaultProps: {
    type: 'text/css',
    style: {}
  },
  styleData: function styleData(config) {
    return {
      all: {
        '.ap-spinner': {
          textAlign: 'center',
          display: 'none'
        },
        '.ap-spinner.ap-spinner-enabled': {
          display: 'inline-flex',
          justifyContent: 'center',
          alignItems: 'center',
          width: '100%',
          height: '100%'
        },
        '.ap-spinner-icon': {
          display: 'inline-block',
          margin: '0 4px',
          transition: 'opacity 100ms',
          opacity: 1
        },
        '.ap-spinner-cover': {
          position: 'absolute',
          left: 0,
          top: 0,
          right: 0,
          bottom: 0,
          zIndex: 4,
          justifyContent: 'center',
          alignItems: 'center',
          background: _apemanReactStyle.ApStyle.COVER_BACKGROUND,
          color: _apemanReactStyle.ApStyle.COVER_COLOR
        }
      },
      small: {},
      medium: {},
      large: {}
    };
  }
});

exports.default = ApSpinnerStyle;
