// apeman-react-spinner/shim/node/ap_spinner.js
/**
 * apeman react package for spinner.
 * @class ApSpinner
 */

'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

var _asobj = require('asobj');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var SpinnerThemes = {
  a: ['fa', 'fa-spin', 'fa-spinner'],
  b: ['fa', 'fa-spin', 'fa-circle-o-notch'],
  c: ['fa', 'fa-spin', 'fa-refresh'],
  d: ['fa', 'fa-spin', 'fa-gear'],
  e: ['fa', 'fa-spin', 'fa-pulse']
};
var SpinnerSizes = ['medium', 'xx-small', 'x-small', 'small', 'large', 'x-large', 'xx-large', 'smaller', 'larger'];

var DEFAULT_THEME = 'c';
var DEFAULT_SIZE = 'large';

/** @lends ApSpinner */

var ApSpinner = function (_Component) {
  _inherits(ApSpinner, _Component);

  function ApSpinner() {
    _classCallCheck(this, ApSpinner);

    return _possibleConstructorReturn(this, Object.getPrototypeOf(ApSpinner).apply(this, arguments));
  }

  _createClass(ApSpinner, [{
    key: 'render',
    value: function render() {
      var s = this;
      var props = s.props;

      var className = (0, _classnames2.default)('ap-spinner', props.className, {
        'ap-spinner-cover': !!props.cover,
        'ap-spinner-enabled': !!props.enabled
      });
      return _react2.default.createElement(
        'div',
        { className: className,
          style: Object.assign({}, props.style) },
        _react2.default.createElement('span', { className: (0, _classnames2.default)('ap-spinner-icon', SpinnerThemes[props.theme]),
          style: { fontSize: props.size }
        })
      );
    }

    // --------------------
    // Lifecycle
    // --------------------

  }, {
    key: 'shouldComponentUpdate',
    value: function shouldComponentUpdate(nextProps, nextState) {
      var s = this;
      var props = s.props;
      var state = s.state;

      return !(0, _asobj.shallowEqual)(props, nextProps) || (0, _asobj.shallowEqual)(state, nextState);
    }
  }, {
    key: 'componentDidMount',
    value: function componentDidMount() {
      var s = this;
      s.setState({
        iconVisible: true
      });
    }
  }, {
    key: 'componentWillUnmount',
    value: function componentWillUnmount() {
      var s = this;
    }

    // --------------------
    // Custom
    // --------------------

  }]);

  return ApSpinner;
}(_react.Component);

Object.assign(ApSpinner, {
  // --------------------
  // Specs
  // --------------------
  propTypes: {
    /** Shows spin */
    enabled: _react.PropTypes.bool,
    /** Theme of the spinner */
    theme: _react.PropTypes.oneOf(Object.keys(SpinnerThemes)),
    /** Size of spinner icon */
    size: _react.PropTypes.oneOf(SpinnerSizes),
    /** show as cover */
    cover: _react.PropTypes.bool
  },
  defaultProps: {
    enabled: false,
    theme: DEFAULT_THEME,
    size: DEFAULT_SIZE
  },
  DEFAULT_THEME: DEFAULT_THEME
});

exports.default = ApSpinner;
