// apeman-react-spinner/shim/node/with_spin.js
/**
 * Higher order component with spinner
 * @function withSpin
 * @param {function} Component - A component constructor
 * @returns {function} - Wrapped component
 */
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _breact = require('breact');

var _asobj = require('asobj');

var _ap_spinner = require('./ap_spinner');

var _ap_spinner2 = _interopRequireDefault(_ap_spinner);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/** @lends withSpin */
function withSpin(Component) {
  var propTypes = {
    /** Show loading spinner */
    spinning: _react.PropTypes.bool,
    /** Theme of the spinner */
    spinnerTheme: _react.PropTypes.string,
    /** Size of spinner */
    spinnerSize: _react.PropTypes.string,
    /** Spinner instance */
    spinner: _react.PropTypes.node
  };
  return (0, _breact.wrap)(Component, {
    displayName: 'withSpin',

    propTypes: propTypes,

    getDefaultProps: function getDefaultProps() {
      return {
        spinning: false,
        spinnerTheme: _ap_spinner2.default.DEFAULT_THEME,
        spinnerSize: 'x-large',
        spinner: undefined
      };
    },
    render: function render() {
      var s = this;
      var props = s.props;
      var spinnerTheme = props.spinnerTheme;
      var spinnerSize = props.spinnerSize;
      var spinning = props.spinning;
      var _props$spinner = props.spinner;
      var spinner = _props$spinner === undefined ? _react2.default.createElement(_ap_spinner2.default, { theme: spinnerTheme,
        className: 'ap-spinner-cover',
        size: spinnerSize,
        enabled: spinning }) : _props$spinner;

      var componentProps = (0, _asobj.clone)(props, {
        without: Object.keys(propTypes)
      });
      return _react2.default.createElement(
        Component,
        _extends({}, componentProps, {
          spinner: spinner
        }),
        props.children
      );
    }
  });
}

exports.default = withSpin;
