// apeman-react-image/shim/node/ap_image_style.js
/**
 * Style for ApImage.
 * @class ApImageStyle
 */

'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _assign = require('babel-runtime/core-js/object/assign');

var _assign2 = _interopRequireDefault(_assign);

var _getPrototypeOf = require('babel-runtime/core-js/object/get-prototype-of');

var _getPrototypeOf2 = _interopRequireDefault(_getPrototypeOf);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _apemanReactStyle = require('apeman-react-style');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/** @lends ApImageStyle */
var ApImageStyle = function (_Component) {
  (0, _inherits3.default)(ApImageStyle, _Component);

  function ApImageStyle() {
    (0, _classCallCheck3.default)(this, ApImageStyle);
    return (0, _possibleConstructorReturn3.default)(this, (ApImageStyle.__proto__ || (0, _getPrototypeOf2.default)(ApImageStyle)).apply(this, arguments));
  }

  (0, _createClass3.default)(ApImageStyle, [{
    key: 'render',
    value: function render() {
      var s = this;
      var props = s.props;

      var _ApImageStyle$styleDa = ApImageStyle.styleData(props);

      var all = _ApImageStyle$styleDa.all;
      var small = _ApImageStyle$styleDa.small;
      var medium = _ApImageStyle$styleDa.medium;
      var large = _ApImageStyle$styleDa.large;


      return _react2.default.createElement(
        _apemanReactStyle.ApStyle,
        { data: (0, _assign2.default)(all, props.style),
          smallMediaData: small,
          mediumMediaData: medium,
          largeMediaData: large
        },
        props.children
      );
    }
  }]);
  return ApImageStyle;
}(_react.Component);

(0, _assign2.default)(ApImageStyle, {
  propTypes: {
    style: _react.PropTypes.object,
    backgroundColor: _react.PropTypes.string
  },

  defaultProps: {
    style: {},
    backgroundColor: '#DDD',
    spinColor: 'rgba(255,255,255,0.5)'
  },
  styleData: function styleData(config) {
    var backgroundColor = config.backgroundColor;
    var spinColor = config.spinColor;

    var transitionDuration = 100;
    return {
      all: {
        '.ap-image': {
          backgroundColor: '' + backgroundColor,
          overflow: 'hidden',
          textAlign: 'center',
          position: 'relative',
          display: 'inline-flex',
          justifyContent: 'center',
          alignItems: 'center'
        },
        '.ap-image img': {
          opacity: 0,
          transition: 'width ' + transitionDuration + 'ms, opacity ' + transitionDuration + 'ms'
        },
        '.ap-image-ready img': {
          opacity: 1
        },
        '.ap-image-content': {
          position: 'relative',
          boxSizing: 'border-box'
        },
        '.ap-image-spinner': {
          position: 'absolute',
          left: 0,
          top: 0,
          right: 0,
          bottom: 0,
          textAlign: 'center',
          display: 'inline-flex',
          zIndex: 8,
          backgroundColor: 'rgba(0,0,0,0.1)',
          color: '' + spinColor,
          justifyContent: 'center',
          alignItems: 'center'
        },
        '.ap-image-notfound': {
          display: 'block',
          textAlign: 'center',
          color: 'rgba(0,0,0,0.1)',
          fontFamily: 'monospace'
        }
      },
      small: {},
      medium: {},
      large: {}
    };
  }
});

exports.default = ApImageStyle;
