// apeman-react-slideshow/lib/ap_slideshow_style.jsx
/**
 * Style for ApSlideshow.
 * @class ApSlideshowStyle
 */

'use strict'

import React, {PropTypes as types} from 'react'
import {ApStyle} from 'apeman-react-style'

/** @lends ApSlideshowStyle */
const ApSlideshowStyle = React.createClass({
  propTypes: {
    style: types.object
  },
  getDefaultProps () {
    return {
      style: {}
    }
  },
  statics: {
    styleData (config) {
      let paddingRate = -25
      return {
        all: {
          '.ap-slideshow': {
            position: 'relative',
            display: 'block',
            overflow: 'hidden'
          },
          '.ap-slideshow-inner': {
            position: 'relative',
            display: 'block'
          },
          '.ap-slideshow .ap-slideshow-image': {
            position: 'absolute',
            display: 'inline-block',
            visibility: 'visible',
            left: `${paddingRate}%`,
            top: `${paddingRate}%`,
            right: `${paddingRate}%`,
            bottom: `${paddingRate}%`,
            boxSizing: 'border-box',
            opacity: 0,
            width: `${100 - (paddingRate * 2)}%`,
            height: `${100 - (paddingRate * 2)}%`
          },
          '.ap-slideshow .ap-image-content': {
            position: 'static'
          },
          '.ap-slideshow .ap-slideshow-image-active': {
            visibility: 'visible',
            opacity: '1 !important',
            transform: 'translate(0, 0) scale(1, 1) skew(0, 0) !important'
          }
        }
      }
    }
  },
  render () {
    const s = this
    let { props } = s

    let { all, small, medium, large } = ApSlideshowStyle.styleData(props)

    return (
      <ApStyle data={ Object.assign(all, props.style) }
               smallMediaData={ small }
               mediumMediaData={ medium }
               largeMediaData={ large }
      >{ props.children }</ApStyle>
    )
  }
})

export default ApSlideshowStyle
