// apeman-react-slideshow/shim/node/ap_slideshow.js
/**
 * apeman react package for slideshow components.
 * @class ApSlideshow
 */

'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

var _image_effect = require('./effects/image_effect');

var _image_effect2 = _interopRequireDefault(_image_effect);

var _apemanReactImage = require('apeman-react-image');

var _apemanReactTouchable = require('apeman-react-touchable');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/** @lends ApSlideshow */
var ApSlideshow = _react2.default.createClass({
  displayName: 'ApSlideshow',


  // --------------------
  // Specs
  // --------------------

  propTypes: {
    /** Image URLs */
    images: _react.PropTypes.array,
    /** Repeat flipping */
    repeat: _react.PropTypes.bool,
    /** Component width */
    width: _react.PropTypes.number,
    /** Component height */
    height: _react.PropTypes.number,
    /** Flip interval */
    interval: _react.PropTypes.number,
    /** Flip effect */
    effect: _react.PropTypes.oneOf(['fadeAndScale'])
  },

  mixins: [],

  statics: {},

  getInitialState: function getInitialState() {
    return {
      index: 0
    };
  },
  getDefaultProps: function getDefaultProps() {
    return {
      images: [],
      repeat: false,
      interval: 20000,
      height: 280,
      width: 420,
      effect: 'fadeAndScale'
    };
  },
  render: function render() {
    var s = this;
    var state = s.state;
    var props = s.props;


    var size = {
      width: props.width,
      height: props.height
    };

    var curIndex = state.index % (props.images.length - 1);
    var images = props.images.map(function (image, i) {
      var prev = i === curIndex - 1;
      var cur = i === curIndex;

      var effect = (0, _image_effect2.default)(props.effect, {
        duration: props.interval
      });

      return s._renderImage(image, {
        'ap-slideshow-image-active': prev || cur
      }, i, effect);
    });

    return _react2.default.createElement(
      'div',
      { className: (0, _classnames2.default)('ap-slideshow', props.className),
        style: Object.assign({}, size, props.style) },
      _react2.default.createElement(
        _apemanReactTouchable.ApTouchable,
        { onTap: s.tick },
        _react2.default.createElement(
          'div',
          { className: 'ap-slideshow-inner',
            style: Object.assign({}, size) },
          images
        )
      )
    );
  },


  // --------------------
  // Lifecycle
  // --------------------

  componentWillMount: function componentWillMount() {
    var s = this;
  },
  componentDidMount: function componentDidMount() {
    var s = this;
    s.startTicking();

    setTimeout(function () {
      s.tick();
    }, 800);
  },
  componentWillUnmount: function componentWillUnmount() {
    var s = this;
    s.stopTicking();
  },


  // ------------------
  // Helper
  // ------------------

  startTicking: function startTicking() {
    var s = this;
    s._ticking = true;
    s.tick();
  },
  stopTicking: function stopTicking() {
    var s = this;
    s._ticking = false;
    clearTimeout(s._tickTimer);
    s._tickTimer = null;
  },
  tick: function tick() {
    var s = this;
    var props = s.props;
    var state = s.state;
    var index = state.index;


    s.setState({
      index: index + 1
    });
    clearTimeout(s._tickTimer);
    s._tickTimer = setTimeout(function () {
      s.tick();
    }, props.interval);
  },


  // ------------------
  // Private
  // ------------------

  _tickTimer: null,
  _renderImage: function _renderImage(src, className, i, style) {
    var s = this;
    return _react2.default.createElement(_apemanReactImage.ApImage, { className: (0, _classnames2.default)('ap-slideshow-image', className),
      style: Object.assign({}, style),
      key: 'ap-slideshow-image-' + i,
      scale: 'fill',
      width: '150%',
      height: '150%',
      src: src });
  }
});

exports.default = ApSlideshow;
