// apeman-react-touchable/shim/node/ap_touchable_style.js
/**
 * Style for ApTouchable.
 * @class ApTouchableStyle
 */

'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _assign = require('babel-runtime/core-js/object/assign');

var _assign2 = _interopRequireDefault(_assign);

var _getPrototypeOf = require('babel-runtime/core-js/object/get-prototype-of');

var _getPrototypeOf2 = _interopRequireDefault(_getPrototypeOf);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _apemanReactStyle = require('apeman-react-style');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/** @lends ApTouchableStyle */
var ApTouchableStyle = function (_Component) {
  (0, _inherits3.default)(ApTouchableStyle, _Component);

  function ApTouchableStyle() {
    (0, _classCallCheck3.default)(this, ApTouchableStyle);
    return (0, _possibleConstructorReturn3.default)(this, (ApTouchableStyle.__proto__ || (0, _getPrototypeOf2.default)(ApTouchableStyle)).apply(this, arguments));
  }

  (0, _createClass3.default)(ApTouchableStyle, [{
    key: 'render',
    value: function render() {
      var s = this;
      var props = s.props;

      var _ApTouchableStyle$sty = ApTouchableStyle.styleData(props);

      var all = _ApTouchableStyle$sty.all;
      var small = _ApTouchableStyle$sty.small;
      var medium = _ApTouchableStyle$sty.medium;
      var large = _ApTouchableStyle$sty.large;


      return _react2.default.createElement(
        _apemanReactStyle.ApStyle,
        { data: (0, _assign2.default)(all, props.style),
          smallMediaData: small,
          mediumMediaData: medium,
          largeMediaData: large
        },
        props.children
      );
    }
  }]);
  return ApTouchableStyle;
}(_react.Component);

(0, _assign2.default)(ApTouchableStyle, {
  propTypes: _apemanReactStyle.ApStyle.propTypes,
  defaultProps: {},
  styleData: function styleData(config) {
    return {
      all: {
        '.ap-touchable': {}
      }
    };
  }
});

exports.default = ApTouchableStyle;
