// apeman-react-touchable/shim/node/ap_touchable.js
/**
 * apeman react package for touchable component.
 * @class ApTouchable
 */

'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ApTouchable = undefined;

var _assign = require('babel-runtime/core-js/object/assign');

var _assign2 = _interopRequireDefault(_assign);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

var _with_touch = require('./with_touch');

var _with_touch2 = _interopRequireDefault(_with_touch);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/** @lends ApTouchable */
var ApTouchable = _react2.default.createClass({
  displayName: 'ApTouchable',


  // --------------------
  // Specs
  // --------------------
  propTypes: {},

  statics: {},

  getInitialState: function getInitialState() {
    return {};
  },
  getDefaultProps: function getDefaultProps() {
    return {};
  },
  render: function render() {
    var s = this;
    var props = s.props;


    return _react2.default.createElement(
      'div',
      { id: props.id,
        className: (0, _classnames2.default)('ap-touchable', props.className),
        style: (0, _assign2.default)({}, props.style) },
      props.children
    );
  }

  // --------------------
  // Lifecycle
  // --------------------

  // ------------------
  // Helper
  // ------------------

  // ------------------
  // Private
  // ------------------

});

exports.ApTouchable = ApTouchable;
exports.default = (0, _with_touch2.default)(ApTouchable);
