// apeman-react-touchable/shim/node/with_outside.js
/**
 * Higher order component with outside touch
 * @function withOutside
 * @param {function} Component - A component constructor
 * @returns {function} - Wrapped component
 */
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _breact = require('breact');

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _reactDom = require('react-dom');

var _reactDom2 = _interopRequireDefault(_reactDom);

var _bwindow = require('bwindow');

var _new_hammer = require('./helpers/new_hammer');

var _new_hammer2 = _interopRequireDefault(_new_hammer);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/** @lends withOutside */
function withOutside(Component, handleOutSide) {
  return (0, _breact.wrap)(Component, {
    propTypes: {
      onOutside: _react.PropTypes.func
    },

    componentDidMount: function componentDidMount() {
      var s = this;
      var body = (0, _bwindow.get)('document.body');
      var hammer = (0, _new_hammer2.default)(body);
      hammer.on('tap', s.handleTapForOutside);
      s.hammer = hammer;
    },
    componentWillUnmount: function componentWillUnmount() {
      var s = this;
      var hammer = s.hammer;

      hammer.off('tap', s.handleTapForOutside);
      hammer.destroy();
      hammer.stop();
      hammer.destroy();
    },
    handleTapForOutside: function handleTapForOutside(e) {
      var s = this;
      var props = s.props;
      var _props$onOutside = props.onOutside;
      var onOutside = _props$onOutside === undefined ? function () {
        return undefined;
      } : _props$onOutside;

      var node = _reactDom2.default.findDOMNode(s);
      if (!node) {
        return;
      }
      var contained = node.contains(e.target);
      if (!contained) {
        onOutside(e);
        if (handleOutSide) {
          handleOutSide(e, { element: s });
        }
      }
    }
  });
}

exports.default = withOutside;
